Add save/load to 6 independent egs mode

lzulauf opened the pull request:
It will now detect slide and pot changes (for enabled channels). Once a
change has occurred, and at least 5 seconds have passed since the last
change, the state will be written to flash memory so that the envelopes
can survive a power cycle.

There is more work to be done around the mode's static variables - they
initialize only once upon startup (to dummy values). This both makes it
detect initial slider positions on startup as moves (if not at zero), as
well as makes it not reset the initial slider positions when changing
modes.

A wav file with the current code can be found at https://github.com/lzulauf/qiemem_stages/actions/runs/13554018901

A separate wav file which lights up all LEDS yellow when there is a pending save and all red when a save occurs can be found here (current code minus the last commit that removed the led debug code):  https://github.com/lzulauf/qiemem_stages/actions/runs/13553530768

We can either merge this as is (knowing that there's still some weirdness upon startup and switching modes) or we can wait until a second PR is ready with those fixes (no ETA currently).

I also wasn't sure the best way to put copyright notices on new files. There's clearly not a pattern of adding names to the copyright of existing files, so I'm not sure what the best pattern to follow here is.

lzulauf committed:
Add save/load to 6 independent egs mode

It will now detect slide and pot changes (for enabled channels). Once a
change has occurred, and at least 5 seconds have passed since the last
change, the state will be written to flash memory so that the envelopes
can survive a power cycle.

There is more work to be done around the mode's static variables - they
initialize only once upon startup (to dummy values). This both makes it
detect initial slider positions on startup as moves (if not at zero), as
well as makes it not reset the initial slider positions when changing
modes.

lzulauf committed:
Remove debug leds and dead code

Saving and loading are _generally_ working.

There is still more to do when switching modes and startup.

There should be a more explicit reinit for the mode that sets all the state and does not rely upon static variables.

lzulauf commented:
The followup PR is ready - it's base branch is currently set to this branch, which is causing the PR to live in my repository for now. After this PR merges I'll reopen the second PR in this repository. Or if you prefer, I can merge that PR into this branch prior to landing this PR (though resulting in a larger diff).

You can preview that second PR here: https://github.com/lzulauf/qiemem_stages/pull/2

lzulauf committed:
Rename manager mode

Also make naming more consistent with followup PR

lzulauf committed:
Cleanup

lzulauf committed:
Push comment upstream

lzulauf committed:
Rename envelope_mode back to envelope_manager

lzulauf committed:
More EnvelopeMode->EnvelopeManager renames

qiemem commented on stages/envelope_manager.cc:
```diff
@@ -0,0 +1,195 @@
+
+  const uint8_t kNoiseTolerance = 1;
+
+  // Convert a pot or slider value on a float [-1., 2.) range to a [0, 255]
+  // range.
+  uint16_t PotOrSliderToUint16(float value) {
+    float result = (value+1.0f) / 3.0f * 255.0f;
+    CONSTRAIN(result, 0, 255);
+    return result;
+  }
```
The range of pots/sliders is [0.0,1.0) (see https://github.com/qiemem/eurorack/blob/1d401208eb011ba0c255e4da8dd3780aa3c9896b/stages/drivers/pots_adc.h#L53-L59; note `values_` is a `uint16_t`). It only becomes [-1.0, 2.0) once you add cv values to the sliders. You want to scale by `256.0f` rather than `255.0f` or else you won't hit `255` even when the adc gives 65535. Also, minor nit, but the `ToUint16`  implies that the range will be [0, 65535] to me. Overflow won't be an issue here with the `CONSTRAIN`, so I don't think there's a reason to return `uint16_t`.

lzulauf replied:
Well that makes _so_ much more sense.

Yeah, that should be a uint8 - I think it was leftover from when the conversion was storing both the pot and slider in a single uint16.

qiemem commented on stages/envelope_manager.cc:
```diff
@@ -0,0 +1,195 @@
+    return result;
+  }
+
+  // Convert a uint16_t on an integer [0, 255] range to a [-1., 2.) range
+  // suitable for a pot or slider.
+  float Uint16ToSliderOrPot(uint16_t value) {
+    return (value / 255.0f * 3.0f) - 1.0f;
+  }
```
See comments for `PotOrSliderToUint16`

qiemem commented on stages/settings.h:
```diff
@@ -72,12 +72,38 @@ struct PersistentData {
+// - byte 9: -- padding --
+// - byte 10: release level
+// - byte 11: release curve
+const uint8_t IEG_DELAY_LENGTH = 0;
+const uint8_t IEG_ATTACK_LENGTH = 2;
+const uint8_t IEG_ATTACK_CURVE = 3;
+const uint8_t IEG_HOLD_LENGTH = 4;
+const uint8_t IEG_DECAY_LENGTH = 6;
+const uint8_t IEG_DECAY_CURVE = 7;
+const uint8_t IEG_SUSTAIN_LEVEL = 8;
+const uint8_t IEG_RELEASE_LENGTH = 10;
+const uint8_t IEG_RELEASE_CURVE = 11;
```
Any reason not to use an enum here? You could do `enum IEGParam : uint8_t` to keep it 1 byte.

qiemem commented on stages/settings.h:
```diff
@@ -72,12 +72,38 @@ struct PersistentData {
   uint16_t segment_configuration[kNumChannels];
   uint8_t color_blind;
   uint8_t multimode;
+  uint8_t independent_eg_state[kNumChannels][12];
```
If you don't want to have to do math, you could also do:

```
uint8_t independent_eg_state[kNumChannels][6][2];
```

Memory layout should be identical.

lzulauf replied:
I might leave this as is for now. If we used pointer offsets using the enum values I think it would get confusing since the offsets don't match up with how the state is declared. If we use a triple nested lookup then we would either get rid of the enums and just use hardcoded indexes (attack slider = [1][0]) or else we would want to also add enum values for pot vs slider (attack slider = [IEG_ATTACK][IEG_SLIDER], and there would be invalid combinations (e.g. [IEG_DELAY][IEG_POT] isn't used).

qiemem commented on stages/stages.cc:
```diff
@@ -277,41 +293,49 @@ void ProcessSixIndependentEgs(IOBuffer::Block* block, size_t size) {
     }
   }
+
+  if (save_timer >= 0) {
+    if (++save_timer >= save_time_wait) {
+      save_timer = -1;
+      settings.SaveState();
+    }
+  }
```
Saving in the main loop like this makes me slightly hesitant. I *think* the main concern is that if a systick interrupt occurs during the save process, you could end up saving inconsistent state or, worst case, the interrupt might try to save while the main loop is already saving... I'm not actually sure what the result would be. Probably a corruption of memory and perhaps a deadlock to go with it. That said, this will be fixed with the generalized debounce save I have in mind. So probably okay to leave as is for now.

lzulauf replied:
Good call - I thought this wasn't the first SaveState occuring in the main thread, but upon re-checking you're absolutely right. If you're comfortable with cleaning this up as part of the debounce, I might take you up on that :)

lzulauf committed:
Address feedback

Fix ranges for slider and pot value storing (should get double the
precision now) and convert IEGParams to an enum.
